### PR TITLE
vscode-json-languageserver support

### PR DESCRIPTION
**Pull request by daliusd**
VSCode JSON languageserver has schema support for linting and
completions.

I have enabled snippets support (`snippetSupport`) even if it is not
fully supported. `label` that comes with completions response can be
used as well.

### CODE REVIEW

**Comment by daliusd** [comment]
Please do not merge yet. I want to investigate some alternatives.

**Comment by daliusd** [comment]
Now everything is good.

**Comment by daliusd** [comment]
@hsanson it is not WIP anymore 😄 Unless you have some ideas what can be improved.

**Comment by daliusd** [comment]
Some examples how it looks like in action:
image

image

Schema json `person.json`:

```json
{
  "$id": "https://example.com/person.schema.json",
  "title": "Person",
  "type": "object",
  "properties": {
    "firstName": {
      "type": "string",
      "description": "The person's first name."
    },
    "lastName": {
      "type": "string",
      "description": "The person's last name."
    },
    "age": {
      "description": "Age in years which must be equal to or greater than zero.",
      "type": "integer",
      "minimum": 0
    }
  }
}
```

Sample test file:
```json
{
  "$schema": "./person.json",
  "firstName": "John",
  "lastName": "Doe",
  "age": 2
}
```

File: autoload/ale/completion.vim [hunk 1]
```diff
@@ -581,9 +581,14 @@ function! ale#completion#ParseLSPCompletions(response) abort
             continue
         endif
 
-        if get(l:item, 'insertTextFormat', s:LSP_INSERT_TEXT_FORMAT_PLAIN) is s:LSP_INSERT_TEXT_FORMAT_PLAIN
+        let l:insertTextFormat = get(l:item, 'insertTextFormat', s:LSP_INSERT_TEXT_FORMAT_PLAIN)
+
+        if l:insertTextFormat is s:LSP_INSERT_TEXT_FORMAT_PLAIN
         \&& type(get(l:item, 'textEdit')) is v:t_dict
             let l:text = l:item.textEdit.newText
+        elseif l:insertTextFormat is s:LSP_INSERT_TEXT_FORMAT_SNIPPET
```
Comment (daliusd) [inline autoload/ale/completion.vim:589]: This is there we have fake snipped support. Still it is better than nothing as `label` can be used for auto-complete.

Comment (daliusd) [reply autoload/ale/completion.vim:589]: It is safe to do this as below we have call to `ale#completion#Filter`.

File: doc/ale.txt [hunk 2]
```diff
@@ -2951,6 +2952,7 @@ documented in additional help files.
     jq....................................|ale-json-jq|
     prettier..............................|ale-json-prettier|
     spectral..............................|ale-json-spectral|
+    vscodejson............................|ale-json-vscode|
```
Comment (hsanson) [inline doc/ale.txt:2955]: Please add an entry for this new tool in the `ale-supported-languages-and-tools.txt` and `supported-tools.md` files.

File: ale_linters/json/vscodejson.vim [hunk 3]
```diff
@@ -0,0 +1,14 @@
+" Author: Dalius Dobravolskas <dalius.dobravolskas@gmail.com>
+" Description: VSCode json languageserver
+
+function! ale_linters#json#vscodejson#GetProjectRoot(buffer) abort
```
Comment (hsanson) [inline ale_linters/json/vscodejson.vim:4]: This is not good idea. This means that for every JSON file open in vim a new vscode-json-languageserver will be spawn.  Better look for generic project files such as the presence of a .git folder to set as project root.

File: autoload/ale/completion.vim (hunk 1 above)
Comment (hsanson) [reply autoload/ale/completion.vim:589]: Not well versed in language servers so will have to ask @w0rp to take a look here.

**Review by hsanson — CHANGES_REQUESTED** [review]

File: ale_linters/json/vscodejson.vim (hunk 3 above)
Comment (daliusd) [reply ale_linters/json/vscodejson.vim:4]: OK. Changed to git root while I think opening it for every JSON is OK as well.

File: doc/ale.txt (hunk 2 above)
Comment (daliusd) [reply doc/ale.txt:2955]: Done.

File: autoload/ale/completion.vim (hunk 1 above)
Comment (daliusd) [reply autoload/ale/completion.vim:589]: OK. Let's see what @w0rp has to say. It solves problem with this LSP but I'm pretty sure other LSPs will benefit accidentally without causing any accidental harm. Proper snippet support would be even better potentially, but that's really big task IMHO.

File: autoload/ale/lsp.vim [hunk 4]
```diff
@@ -412,7 +412,7 @@ function! s:SendInitMessage(conn) abort
     \               'completion': {
     \                   'dynamicRegistration': v:false,
     \                   'completionItem': {
-    \                       'snippetSupport': v:false,
+    \                       'snippetSupport': v:true,
```
Comment (hsanson) [inline autoload/ale/lsp.vim:415]: Unfortunatelly I think this needs more evaluation.  I did some tests and when this is enabled ALEImport breaks for me.
If enablinb snippetSupport is not critical for vscodejson I would recommend reverting this to get vscodejson merged and then handle adding snippetSupport on a separate issue.

Comment (daliusd) [reply autoload/ale/lsp.vim:415]: Absolutely not critical. Let me fix it.

File: autoload/ale/completion.vim (hunk 1 above)
Comment (daliusd) [reply autoload/ale/completion.vim]: Removed from this PR. Let's do this as separate PR.

**Review by hsanson — APPROVED** [review]
Thanks, looks good.

File: autoload/ale/lsp.vim (hunk 4 above)
Comment (daliusd) [reply autoload/ale/lsp.vim]: @hsanson BTW, how ALEImport breaks for you? What filetype you were workin on? I want to reproduce the problem and see if there is a fix.

Comment (hsanson) [reply autoload/ale/lsp.vim]: Simple Java HelloWorld.java managed via Gradle:

```
public class HelloWorld {
    public static void main(String[] argv) {
        LocalDate today = LocalDate.now();
        System.out.println("Hello gradle world at " + today);
    }
}
```

1. Put the cursos in LocalDate class and invoke :ALEImport.
2. ALE should add the import statement "import java.time.LocalDate" at the beginning of the file.
3. With the snippetSupport enabled this does not work.

**Pull request merged**